crates/gpui/src/platform/windows/single_instance.rs

```diff
@@ -0,0 +1,124 @@
+use parking_lot::RwLock;
+use util::ResultExt;
+use windows::Win32::{
+    Foundation::{CloseHandle, GetLastError, ERROR_ALREADY_EXISTS, MAX_PATH},
+    System::{
+        Memory::{MapViewOfFile, OpenFileMappingW, UnmapViewOfFile, FILE_MAP_WRITE},
+        Threading::{CreateEventW, OpenEventW, SetEvent, EVENT_MODIFY_STATE},
+    },
+};
+use windows_core::HSTRING;
+
+use super::APP_SHARED_MEMORY_MAX_SIZE;
+
+static APP_IDENTIFIER: RwLock<String> = RwLock::new(String::new());
+static APP_INSTANCE_EVENT_IDENTIFIER: RwLock<String> = RwLock::new(String::new());
+static APP_SHARED_MEMORY_IDENTIFIER: RwLock<String> = RwLock::new(String::new());
+
+pub(crate) fn register_app_identifier(app_identifier: &str, local: bool) {
+    if app_identifier.len() as u32 > MAX_PATH {
+        panic!(
+            "The length of app identifier `{app_identifier}` is limited to {MAX_PATH} characters."
+        );
+    }
+    *APP_IDENTIFIER.write() = app_identifier.to_string();
+    let (sync_event_identifier, shared_memory_identifier) = if local {
+        (
+            format!("Local\\{app_identifier}-Instance-Event"),
+            format!("Local\\{app_identifier}-Shared-Memory"),
+        )
+    } else {
+        (
+            format!("Global\\{app_identifier}-Instance-Event"),
+            format!("Global\\{app_identifier}-Shared-Memory"),
+        )
+    };
+    if sync_event_identifier.len() as u32 > MAX_PATH {
+        panic!("The length of app identifier `{sync_event_identifier}` is limited to {MAX_PATH} characters.");
+    }
+    *APP_INSTANCE_EVENT_IDENTIFIER.write() = sync_event_identifier;
+    if shared_memory_identifier.len() as u32 > MAX_PATH {
+        panic!("The length of app identifier `{shared_memory_identifier}` is limited to {MAX_PATH} characters.");
+    }
+    *APP_SHARED_MEMORY_IDENTIFIER.write() = shared_memory_identifier;
+}
+
+pub(crate) fn retrieve_app_identifier() -> String {
+    let lock = APP_IDENTIFIER.read();
+    if lock.is_empty() {
+        panic!("Make sure you have called `register_app_identifier` first.");
+    }
+    lock.to_string()
+}
+
+pub(crate) fn retrieve_app_instance_event_identifier() -> String {
+    let lock = APP_INSTANCE_EVENT_IDENTIFIER.read();
+    if lock.is_empty() {
+        panic!("Make sure you have called `register_app_identifier` first.");
+    }
+    lock.to_string()
+}
+
+pub(crate) fn retrieve_app_shared_memory_identifier() -> String {
+    let lock = APP_SHARED_MEMORY_IDENTIFIER.read();
+    if lock.is_empty() {
+        panic!("Make sure you have called `register_app_identifier` first.");
+    }
+    lock.to_string()
+}
+
+pub(crate) fn check_single_instance<F>(f: F) -> bool
+where
+    F: FnOnce(bool) -> bool,
+{
+    unsafe {
+        CreateEventW(
+            None,
+            false,
+            false,
+            &HSTRING::from(retrieve_app_instance_event_identifier()),
+        )
+        .expect("Unable to create instance sync event")
```

Comment:
The name (the `HSTRING` here) provided when calling `CreateEventW` here must match the name used in the `OpenEventW` call mentioned in the previous comment.